Remove attribute dac_override from type mcprepare

bootlessxfly opened the pull request:
mcprepare violates a neverallow rule around dac_override_allow.
mcprepare simpily needs to be added to dac_override_allow. The
issue comes in that type mcprepare is not loaded untill later in
the sepolicycheck test. This causes the build to fail with a type
not found error when adding mcprepare to dac_override_allow. To
get around this, dac_override_allow has been added to mcprepare.te
after the declaration of types. I added mcprepare to this custom
dac_override_allow to add it as an exception. Then, I added
the original dac_override newverallow rule so that the check
included mcprepare. Inside of system/sepolicy/public/domain.te,
the neverallow policy for dac_overide needs to be removed so that
it is reran once mcprepare is recognized. This allows for the check to
pass without being disbaled.

While this does require changes to system/sepolicy, this would need to
be done regardless as mcprepare would need to be added to domain.te via
dac_override_allow.

Ideally, this could have been set and overridden inside of
vendor/maruos. I tried and was unable to do so.

Signed-off-by: Chris White <bootlessxfly@gmail.com>

bootlessxfly commented:
There is no maru project for system/sepolicy. It may be a good idea to create one to keep up with changes that can not be handled by vendor/maruos/sepolicy.

Here is the corresponding system/sepolicy branch that corresponds to this fix:
https://github.com/bootlessxfly/android_system_sepolicy/tree/mcprepare_dac_override

utzcoz commented on sepolicy/mcprepare.te:
```diff
@@ -2,6 +2,33 @@
+# We need to add mcprepare to the dac_overrise_allowed, but we must do so after
+# type mcprepare is declared.
+
+define(`dac_override_allowed', `{
```
Looks like it is the same as https://cs.android.com/android/platform/superproject/+/master:system/sepolicy/private/domain.te;l=342-364?q=dac_override_allowed. Will `dac_override_allowed` defined here override `dac_override_allowed` at system/sepolicy? Is it permitted to use different name than `dac_override_allowed`, for example `dac_override_allowed_maru`? If we cant' use other name, do we need update it when upgrading Android base version?

bootlessxfly replied:
> Will dac_override_allowed defined here override dac_override_allowed at system/sepolicy?

It will not override it unfortunately. dac_override_allowed is first initialized in system/sepolicy without mcprepare and then all of the normal checks are run (including the failing one). After that dac_override_allowed is re-initialized with mcprepare added and line 30 validates the neverallow rule against dac_override with mcprepare added as an exception.

- If a device is already passing this check in system/policy, then it will also pass this check.

- If a device is failing the check in system/sepolicy, then the maintainer can go and remove the check from system/sepolicy and the rule is still enforced. We could even add a sed command automatically removing the neverallow check to our build process or run post repo sync.

>  Is it permitted to use different name than dac_override_allowed, for example dac_override_allowed_maru?

We can use a different name. This would also probably make this more readable. We would just modify the neverallow rule to use the new variable. I can update the code with this change.

> If we cant' use other name, do we need update it when upgrading Android base version?

Redgardless of whether we use a different name or not, base Android upgrades could possibly cause us to upgrade this. If dac_override_allowed has new types added to it, then dac_override_allowed_maru would need to also need those types added to it. 
I could go in and add comments explaining this to the code in the form of a TODO so we know in future up-porting to check this variable for changes.



Anyways, if these changes sound good to you, I can add them to the PR.

bootlessxfly replied:
I didnt see your comment over at https://github.com/maruos/maruos/issues/154#issuecomment-961193695 before posting the last reply. I see your point. That is giving more power than I realized. Let me do some additional reading on this.

Edit:
It looks like we may be able to set up a dontaudit rule for this instead. Ill play around with this some in the next couple days.

Here is the avc:

>libsepol.report_failure: neverallow on line 1385 of system/sepolicy/public/domain.te (or line 11832 of policy.conf) violated by allow mcprepare mcprepare:capability { dac_override };
libsepol.check_assertions: 1 neverallow failures occurred
Error while expanding policy
/var/maru/devel/maru-0.7-upstream/out/host/linux-x86/bin/checkpolicy:  loading policy configuration from /var/maru/devel/maru-0.7-upstream/out/target/product/taimen/obj/ETC/sepolicy_neverallows_intermediates/policy.conf

bootlessxfly replied:
I just ran a test where I removed dac_override from mcprepare and reran the build without this change. This got me a clean build and past the dac_override check. I booted it up, launched the container, and connected via wireless display. And everything seems to be working. This seems like the way to go. Remove mcprepare from using dac_override in the first place. I can update this PR with this fix if you want to give it a try.

Edit: I updated this PR to only remove attribute dac_override from type mcprepare

utzcoz commented:
> There is no maru project for system/sepolicy. It may be a good idea to create one to keep up with changes that can not be handled by vendor/maruos/sepolicy.

The reason that MaruOS doesn't folk and create system/sepolicy is to reduce the maintaining problem. When we moved most of MaruOS to app or other  upper level, we can recheck those MaruOS sepolicy. Maybe we can remove some of them.

bootlessxfly commented:
> > There is no maru project for system/sepolicy. It may be a good idea to create one to keep up with changes that can not be handled by vendor/maruos/sepolicy.
> 
> The reason that MaruOS doesn't folk and create system/sepolicy is to reduce the maintaining problem. When we moved most of MaruOS to app or other upper level, we can recheck those MaruOS sepolicy. Maybe we can remove some of them.

That makes sense. An additional project to maintain is probably more hassle than its worth. I mentioned above that we may be able to use something like sed to automatically remove that line whenever we run "repo sync". So I see your point about forking off system/sepolicy.

bootlessxfly committed:
Remove attribute dac_override from type mcprepare

dac_override has a strict neverallow rule keeping only a small list of
allowed types. At first, mcprepare was added to this list. It was
determned that giving mcprepare access to dac_overrde would give the
mcprepare process to much control as it would be able to rwx on all
files. dac_override is an override fordiscretionary access control and
can give non root uid/gids access to any file.

I ran a test to see if maru really needed dac_override, and after
booting things seem to work well. The container launches and connects to
the display with out issues.

If this ends up causing problems, we will need to look into either
setting an dontaudit rule or changing certain file permissions. Since
things seem to work without, lets try completeley removing it.

Signed-off-by: Chris White <bootlessxfly@gmail.com>

utzcoz commented:
Thanks. I will test it on my pixel phones this weekend.

utzcoz reviewed (approved):
LGTM. The device(Pixel XL and Pixel 3a) shows denied log of mcprepare to  access `dac_override` after applying this PR. But `mcprepare`'s function doesn't be affected after removing `dac_override` from type `mcprepare`. I think it is safe to remove it to fix other device's problem.

utzcoz commented:
Hi @pdsouza @pintaf , maybe this PR needs your double-checking on your devices.

bootlessxfly commented:
> LGTM. The device(Pixel XL and Pixel 3a) shows denied log of mcprepare to access `dac_override` after applying this PR. But `mcprepare`'s function doesn't be affected after removing `dac_override` from type `mcprepare`. I think it is safe to remove it to fix other device's problem.

Im still new to the idea of a "dontaudit" rule, but we may do something similar to what was done with mknod with dacoverride. If I am understanding it correctly, this would get rid of the "denied log" messages.

pdsouza commented:
Nice, I'll try this out this week and check

pdsouza commented:
Had some time this evening and ran a build on my Nexus 5. I see the denied logs for `mcprepare` for `dac_override` but it does not impact mcprepare's function. I think this is likely because we have `mcprepare` set to [permissive](https://github.com/maruos/vendor_maruos/blob/maru-0.7/sepolicy/mcprepare.te#L5) for now. 

Anyways, LGTM!

utzcoz commented:
@bootlessxfly Thanks for your work.
